changelog/generate.go

```diff
@@ -0,0 +1,287 @@
+		log.Fatalf("failed to update CHANGELOG: %v", err)
+	}
+
+	if err := UpdateMigrationGuide(Config{
+		File:    migrationFile,
+		Title:   title,
+		Entries: entries,
+	}); err != nil {
+		log.Fatalf("failed to update migration guide: %v", err)
+	}
+}
+
+func LoadEntries(fragmentsDir string) ([]Entry, error) {
+	files, err := ioutil.ReadDir(fragmentsDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read fragments directory: %w", err)
+	}
+
+	var entries []Entry
+	for _, fragFile := range files {
+		if fragFile.Name() == "00-template.yaml" {
+			continue
+		}
+		if fragFile.IsDir() {
+			log.Warnf("Skipping directory %q", fragFile.Name())
+			continue
+		}
+		if filepath.Ext(fragFile.Name()) != ".yaml" || fragFile.IsDir() {
+			log.Warnf("Skipping non-YAML file %q", fragFile.Name())
+			continue
+		}
+		path := filepath.Join(fragmentsDir, fragFile.Name())
+		f, err := os.Open(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open fragment file %q: %w", fragFile.Name(), err)
+		}
+
+		decoder := yaml.NewDecoder(f)
+		fragment := Fragment{}
+		if err := decoder.Decode(&fragment); err != nil {
+			return nil, fmt.Errorf("failed to parse fragment file %q: %w", fragFile.Name(), err)
+		}
+
+		if err := fragment.Validate(); err != nil {
+			return nil, fmt.Errorf("failed to validate fragment file %q: %w", fragFile.Name(), err)
+		}
+
+		entries = append(entries, fragment.Entries...)
+	}
+	return entries, nil
+}
+
+func UpdateChangelog(c Config) error {
+	changelog := map[EntryKind][]string{}
+	for _, e := range c.Entries {
+		changelog[e.Kind] = append(changelog[e.Kind], e.ToChangelogString())
+	}
+
+	var bb bytes.Buffer
+	order := []EntryKind{
+		Addition,
+		Change,
+		Removal,
+		Deprecation,
+		Bugfix,
+	}
+	bb.WriteString(fmt.Sprintf("## %s\n\n", c.Title))
+	for _, k := range order {
+		if entries, ok := changelog[k]; ok {
+			bb.WriteString(k.ToHeader() + "\n\n")
+			for _, e := range entries {
+				bb.WriteString(e + "\n")
+			}
+			bb.WriteString("\n")
+		}
+	}
+
+	existingFile, err := ioutil.ReadFile(c.File)
+	if err != nil {
+		return fmt.Errorf("could not read CHANGELOG: %v", err)
+	}
+	bb.Write(existingFile)
+
+	if err := ioutil.WriteFile(c.File, bb.Bytes(), 0644); err != nil {
+		return fmt.Errorf("could not write CHANGELOG file: %v", err)
+	}
+	return nil
+}
+
+func UpdateMigrationGuide(c Config) error {
+	var bb bytes.Buffer
+	existingFile, err := ioutil.ReadFile(c.File)
+	if err != nil {
+		return fmt.Errorf("could not read migration guide: %v", err)
+	}
+	bb.Write(bytes.Trim(existingFile, "\n"))
+
+	bb.WriteString(fmt.Sprintf("\n\n## %s\n\n", c.Title))
+	haveMigrations := false
+	for _, e := range c.Entries {
+		if e.Migration != nil {
+			haveMigrations = true
+			bb.WriteString(fmt.Sprintf("### %s\n\n", e.Migration.Header))
+			bb.WriteString(fmt.Sprintf("%s\n\n", strings.Trim(e.Migration.Body, "\n")))
+			bb.WriteString(fmt.Sprintf("See %s for more details.\n\n", e.PullRequestLink()))
+		}
+	}
```

Comment:
Do we want to arrange these under headers based on `kind` like we do in the CHANGELOG? Or just flatten the migration entries out and list them one-by-one in the order they're read from the files.

On a separate but related note, do we want to do any sorting of CHANGELOG and migration guide entries?

Reply:
Do we do any sorting of the migration guide now? I've just always added stuff at the end anyway. Though we may want to consider putting the migration guide into version specific upgrade guides when we build, since I think it's starting to get a bit unwieldy. Might be out of scope although if we do want to go that way may be better to do it now than have to rework the tooling later

Reply:
> Do we do any sorting of the migration guide now? 

No, but depending on how we name the fragment files, entries will show up semi-randomly, just based on a lexicographical sort of the filenames.

> Though we may want to consider putting the migration guide into version specific upgrade guides when we build, since I think it's starting to get a bit unwieldy.

Yeah I had the same thought. Scrolling to the bottom of a very long migration guide is not the best UX. We could _at least_ reverse the order so that new migration steps are always at the top.

Reply:
yeah, but if you're two or three versions behind then you need to go to the middle of the document somewhere to find your upgrade instructions, which I think is still just a worse experience than seeing `migration_guide-v0.16.md` or whatever naming scheme we'd use

Reply:
I'm good with this change. And it actually makes the generation simpler since there's no need to deal with an existing file.

Reply:
Migration guides will now be generated in a separate file for each release. In a follow-up, we can split the existing migration guide into separate files per release.